### ISSUE TITLE
Multi-storage

### ISSUE THREAD

**Issue by mjesun** [lukechilds/keyv#49]
Hi!

We are planning to use `keyv` in [`metro`](https://github.com/facebook/metro) as a multi-layered cache system. The goal is to let people mix and match different cache layers, by using the `Map` compatible interface (something we wanted even before finding `keyv`!).

In order to be layered, we need to be able to broadcast key/value combinations from one layer to the rest (essentially, a successful `get` in one layer has to be `set` in all the rest of the layers). However, `keyv` only supports one single adapter per instance.

While the desired approach can be done either as an adapter that proxies to multiple stores, or as a layer built on top of many `keyv` instances, I was curious to know if you would like to have such behavior built inside `keyv`. The current API would not change, but it would allow to pass multiple stores (my idea is to add `stores`, and keep `store` as backward-compatible, which would just re-wrap as `stores = [store]`). Stores would then sequentially be called until we find a positive result.

I'd also like to add a `@keyv/memcached`, as well as a `@keyv/map` (which essentially would be `module.exports = Map`, but it would let combine as well the memory interface by default). For all of that maybe it'd make sense to have a GH `keyv` organization itself (just suggesting).

We'd like also add Flow typing to it (I've seen #22 for TS support), so I guess that one would be OK as well :)

Let me know what you think! 🙂

**Comment by mjesun** [comment]
The broadcasting mechanism would be implemented in the `keyv` layer, not in the adapters, meaning that you could still use a raw `Map` as an adapter. When the upper layer sees a `set`, it will iterate over the stores and perform a get on each of those.

Regarding the consistency, I agree with you. My take on that was return a `Promise.race`, but it is true that not everybody would be comfortable with this (some of them would like to wait for all, or a N/2 quorum, or another strategy). So probably that's a good reason to implement it outside.

The reason for `@keyv/map` is because it would allow us to pass strategies as JSON, by name. I agree this might be controversial too, but given the zero-effort to add it, I just wanted to throw the idea and see what happened. 😀

When it comes to Flow, we can skip adding annotations inside the code and [put them aside instead](https://flow.org/en/docs/libdefs/creation/).

**Comment by lukechilds** [comment]
>The broadcasting mechanism would be implemented in the keyv layer, not in the adapters, meaning that you could still use a raw `Map` as an adapter.

But that only works for a single node right? Imagine you have a web service with two nodes with round robin load balancing and they have a `Map` and Redis store.

```
node1
  Map: {}
  Redis: {}

node2
  Map: {}
  Redis: {}
```

No a user connects to `node1`, and updates some data, the Keyv instance in `node1` will forward this to all other adapters:

```
node1
  Map: { foo: 'bar' }
  Redis: { foo: 'bar' }

node2
  Map: {}
  Redis: { foo: 'bar' }
```

Notice how `node2`s in-memory `Map` adapter is now out of sync, which is the most likely to win in a first to respond race.

It seems like a very hard problem to solve, probably not possible without some P2P communication going on between the Keyv instances. Maybe it would just be easier to document that you shouldn't use an in-memory adapter with other stores if you have multiple nodes.

>The reason for `@keyv/map` is because it would allow us to pass strategies as JSON, by name. I agree this might be controversial too, but given the zero-effort to add it, I just wanted to throw the idea and see what happened. 😀

Yeah, that makes sense.

>When it comes to Flow, we can skip adding annotations inside the code and put them aside instead.

Is there a benefit to Flow over TypeScript. (I'm aware you're probably going to be quite biased towards Flow but fire away 😆). I don't have much experience with either but from what I understand TypeScript has much better IDE autocomplete functionality.

**Comment by mjesun** [comment]
Regarding the example with `Map` + Redis, you are right; but to me it looks an inherent issue of `Map`, rather than the multi-layer system; given the fact that `Map` is the only one to not be in a centralized server. In other words, a combination of `@keyv/memcached`, `@keyv/redis` and `@keyv/sqlite` would stay consistent, since the instance backing them is unique.

Regarding Flow, you are right about the bias 😄. Metro is flow-typed and it would benefit us having them in Flow. That said, Flow types can co-exist in their own folder together with the TypeScript ones. My suggestion is not to type the code itself, but rather provide types in a separate file.

**Comment by lukechilds** [comment]
>Regarding the example with `Map` + Redis, you are right; but to me it looks an inherent issue of `Map`

Yeah, I kinda realised that as I was typing it out. You'd have the same consistency issues if you were using `Map` alone, without any other adapters, so it's not really related to the multi adapter side.

Although it would still probably be worth warning users not to use `Map` in the multi-storage adapter readme just to avoid potential confusion.

>Flow types can co-exist in their own folder together with the TypeScript ones.

Sounds good to me 👍

Will people from the `metro` team be contributing these improvements or are you asking me to implement them? Most of these are features I'd like to add to Keyv anyway but I'm really tied up on paid work atm so the amount of time I can currently put in to OSS is limited.

I'm expecting to have a lot more OSS time after March.

**Comment by mjesun** [comment]
Hi! We would take care of the changes. For the moment, we're working in implementing this inside Metro, but so far with a `keyv` compatible interface. Once we have it and we proved it works, we can make a bunch of PRs to move it to `keyv` itself 🙂

**Comment by lukechilds** [comment]
Awesome, sounds good!

I may start implementing some of these in the mean time if I get time.

**Comment by lukechilds** [comment]
One idea I had regarding type definitions. We could annotate the source with JSDoc comments (or similar).

From this we can generate detailed HTML document which would be great as Keyv seems to be getting quite a bit of interests and the current readme docs are limited by what I can do with markdown.

As well as docs, the comments will contain all type information so it should be possible to generate TypeScript/Flow definitions from the documentation metadata. I don't know if something like this already exists or if we'd have to create it, but it should definitely be possible.

This should also help solve the problem of type definitions getting out of sync with the source. Type definitions could accidentally be left outdated during api changes, however it's much less likely that someone would forget to update some comment metadata that's right above the method they're modifying.

What do you think?

**Comment by lukechilds** [comment]
Two interesting repos, although both seem to be experimental:

https://github.com/englercj/tsd-jsdoc
https://github.com/joarwilk/flowgen

**Comment by Feverqwe** [comment]
Hello! I create simple memjs adapter https://github.com/Feverqwe/keyv-memjs
Maybe someone will need it.
